crates/node/src/providers/robust_ecdsa/presign.rs

```diff
@@ -0,0 +1,294 @@
+use crate::{metrics, tracking};
+use mpc_contract::primitives::domain::DomainId;
+use near_time::Clock;
+use rand::rngs::OsRng;
+use serde::{Deserialize, Serialize};
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+use std::time::Duration;
+use threshold_signatures::ecdsa::robust_ecdsa::{
+    presign::presign, PresignArguments, PresignOutput,
+};
+use threshold_signatures::participants::Participant;
+
+#[derive(derive_more::Deref)]
+pub struct PresignatureStorage(DistributedAssetStorage<PresignOutputWithParticipants>);
+
+// TODO(#1680): simplify alive_participant_ids_query parameter type
+impl PresignatureStorage {
+    pub fn new(
+        clock: Clock,
+        db: Arc<SecretDB>,
+        my_participant_id: ParticipantId,
+        alive_participant_ids_query: Arc<dyn Fn() -> Vec<ParticipantId> + Send + Sync>,
+        domain_id: DomainId,
+    ) -> anyhow::Result<Self> {
+        Ok(Self(DistributedAssetStorage::<
+            PresignOutputWithParticipants,
+        >::new(
+            clock,
+            db,
+            crate::db::DBCol::Presignature,
+            Some(domain_id),
+            my_participant_id,
+            |participants, presignature| {
+                presignature.is_subset_of_active_participants(participants)
+            },
+            alive_participant_ids_query,
+        )?))
+    }
+}
+
+/// Continuously generates presignatures, trying to maintain the desired number of
+/// presignatures available, using the desired number of concurrent computations as
+/// specified in the config.
+///
+/// Note that this function does not take care of the passive side of presignature
+/// generation (i.e. other participants of the computations this function initiates),
+/// so that needs to be separately handled.
+pub(super) async fn run_background_presignature_generation(
+    client: Arc<MeshNetworkClient>,
+    mpc_config: Arc<MpcConfig>,
+    config: Arc<PresignatureConfig>,
+    domain_id: DomainId,
+    presignature_store: Arc<PresignatureStorage>,
+    keygen_out: KeygenOutput,
+) -> anyhow::Result<()> {
+    let in_flight_generations = InFlightGenerationTracker::new();
+    let progress_tracker = Arc::new(PresignatureGenerationProgressTracker {
+        desired_presignatures_to_buffer: config.desired_presignatures_to_buffer,
+        presignature_store: presignature_store.clone(),
+        in_flight_generations: in_flight_generations.num_in_flight_atomic(),
+    });
+    let parallelism_limiter = Arc::new(tokio::sync::Semaphore::new(config.concurrency));
+    let mut tasks = AutoAbortTaskCollection::new();
+
+    let running_participants: Vec<_> = mpc_config
+        .participants
+        .participants
+        .iter()
+        .map(|p| p.id)
+        .collect();
+
+    let threshold = mpc_config.participants.threshold as usize;
+    let num_signers = get_number_of_signers(threshold, running_participants.len());
+    let robust_ecdsa_threshold = translate_threshold(threshold, running_participants.len())?;
```

Comment:
Like, I am suggesting that we error earlier. Here seems a bit late. (This is up for discussion of course and out of scope for this PR).